src/small.rs

```diff
@@ -484,18 +525,132 @@ impl MPCSmall {
+              contributions (544 bytes per PublicKey)
+        */
+
+        // Non-raw sizes.
+        let g1_size = G1Affine::uncompressed_size() as u64; // 96 bytes
+        let g2_size = G2Affine::uncompressed_size() as u64; // 192 bytes
+
+        let mut file = File::open(large_path)?;
+
+        // Read delta_g1, delta_g2, and ic's length.
+        let delta_g1_offset = g1_size + g1_size + g2_size + g2_size; // vk.alpha_g1 + vk.beta_g1 + vk.beta_g2 + vk.gamma_g2
+        file.seek(SeekFrom::Start(delta_g1_offset)).unwrap();
+        let delta_g1 = read_g1(&mut file, check_subgroup)?;
+        let delta_g2 = read_g2(&mut file, check_subgroup)?;
+        let ic_len = file.read_u32::<BigEndian>()? as u64;
+
+        // Read h's length.
+        let h_len_offset = delta_g1_offset + g1_size + g2_size + 4 + ic_len * g1_size; // + vk.delta_g1 + vk.delta_g2 + ic length + ic
+        file.seek(SeekFrom::Start(h_len_offset)).unwrap();
+        let h_len = file.read_u32::<BigEndian>()? as u64;
+
+        // Read l's length.
+        let l_len_offset = h_len_offset + 4 + h_len * g1_size; // + h length + h
+        file.seek(SeekFrom::Start(l_len_offset)).unwrap();
+        let l_len = file.read_u32::<BigEndian>()? as u64;
+
+        // Read a's length.
+        let a_len_offset = l_len_offset + 4 + l_len * g1_size; // + l length + l
+        file.seek(SeekFrom::Start(a_len_offset)).unwrap();
+        let a_len = file.read_u32::<BigEndian>()? as u64;
+
+        // Read b_g1's length.
+        let b_g1_len_offset = a_len_offset + 4 + a_len * g1_size; // + a length + a
+        file.seek(SeekFrom::Start(b_g1_len_offset)).unwrap();
+        let b_g1_len = file.read_u32::<BigEndian>()? as u64;
+
+        // Read b_g2's length.
+        let b_g2_len_offset = b_g1_len_offset + 4 + b_g1_len * g1_size; // + b_g1 length + b_g1
+        file.seek(SeekFrom::Start(b_g2_len_offset)).unwrap();
+        let b_g2_len = file.read_u32::<BigEndian>()? as u64;
+
+        // Read cs_hash.
+        let cs_hash_offset = b_g2_len_offset + 4 + b_g2_len * g2_size; // + b_g2 length + b_g2
+        file.seek(SeekFrom::Start(cs_hash_offset)).unwrap();
+        let mut cs_hash = [0u8; 64];
+        file.read_exact(&mut cs_hash)?;
+
+        // Read contribution's length.
+        let contributions_len = file.read_u32::<BigEndian>()? as u64;
+
+        // Reset seek position.
+        drop(file);
+
+        // Read the (potentially large) h, l, and contributions arrays using buffered io.
+        let file = File::open(large_path)?;
+        let mut reader = BufReader::with_capacity(1024 * 1024, file);
+
+        // Read h.
+        let h_offset = h_len_offset + 4; // + h length
+        reader.seek(SeekFrom::Start(h_offset)).unwrap();
+        let mut h = Vec::<G1Affine>::with_capacity(h_len as usize);
+        for _ in 0..h_len {
+            h.push(read_g1(&mut reader, check_subgroup)?);
+        }
+
+        // Read l. Skip l's length because it was already read.
```

Comment:
skip `h's` lenght?

Reply:
I think that this is correct because the `l` vector proceeds the `h` vector, and I skip over `h`'s length straight to the `h` vector a few lines before this:
```rust
let h_offset = h_len_offset + 4;
```
where `4` is the number of bytes written when serializing `h.len()` as a `u32`.